### PR TITLE
[UI] Add support to download, publish, unpublish, and clone wasm support

### PR DESCRIPTION
**Pull request by Chadha93**
Signed-off-by: Gaurav Chadha <gauravchadha1676@gmail.com>
**Notes for Reviewers**
Add support to download, publish, unpublish, and clone wasm support.

https://github.com/meshery/meshery/assets/65453826/d30f294b-7591-4ac1-96a0-239e20d19c27

This PR fixes #CU-863h1z6br

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by l5io** [comment]
Task linked: [CU-863h1z6br WASM Filters](https://app.clickup.com/t/863h1z6br)

**Comment by Chadha93** [comment]
@Abhishek-kumar09 @theBeginner86, all the filters are having the visibility as private...

**Comment by theBeginner86** [comment]
> @Abhishek-kumar09 @theBeginner86, all the filters are having the visibility as private...

Yes that's true. Its because we haven't published any Filters to catalog so far.

**Comment by leecalcote** [comment]
@aabidsofi19 what comments do you have here?

File: ui/utils/fileDownloader.js [hunk 1]
```diff
@@ -12,6 +12,10 @@ export default function downloadFile({ id, type,name, source_type }) {
     dataUri = `/api/pattern/download/${id}`;
   }
 
+  if (type === "filter") {
```
Comment (aabidsofi19) [inline ui/utils/fileDownloader.js:15]: As we are having a bunch of these if statements , we might try using a object to map type to path  , also having source_type and type together make the function little complex , hence maybe be prone to bugs ( something to consider)

File: ui/components/Filters.js [hunk 2]
```diff
@@ -224,9 +234,39 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
...
+      credentials : "include",
+    },
+    (result) => {
+      if (result) {
+        const capabilitiesRegistry = result;
+        const filtersCatalogueCapability = capabilitiesRegistry?.capabilities.filter((val) => val.feature === MesheryFiltersCatalog);
+        if (filtersCatalogueCapability?.length > 0) setCanPublishFilter(true);
+      }
+    },
+    (err) => console.error(err)
```
Comment (aabidsofi19) [inline ui/components/Filters.js]: can we do something more with the error .

File: ui/components/Filters.js [hunk 3]
```diff
@@ -224,9 +234,39 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
     CLONE_FILTERS : {
       name : "CLONE_FILTER",
       error_msg : "Failed to clone filter file"
+    },
+    PUBLISH_CATALOG : {
+      name : "PUBLISH_CATALOG",
+      error_msg : "Failed to publish catalog"
```
Comment (leecalcote) [inline ui/components/Filters.js:240]: What does this mean? "Failed to publish catalog"

Please rephrase.

Comment (leecalcote) [reply ui/components/Filters.js:240]: Rephrase. Include filter name or filename as confirmation for the user.

File: ui/components/Filters.js [hunk 4]
```diff
@@ -224,9 +234,39 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
...
+    },
+    UNPUBLISH_CATALOG : {
+      name : "PUBLISH_CATALOG",
+      error_msg : "Failed to publish catalog"
```
Comment (leecalcote) [inline ui/components/Filters.js:244]: Rephrase. Include filter name or filename as confirmation for the user.

File: ui/components/Filters.js (hunk 2 above)
Comment (leecalcote) [reply ui/components/Filters.js]: Yes, please, @Chadha93. A toaster?

File: ui/components/Filters.js [hunk 5]
```diff
@@ -241,6 +281,58 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
...
+          options : ["Yes", "No"]
+        });
+        if (response === "Yes") {
+          updateProgress({ showProgress : true });
+          dataFetch(
+                `/api/filter/catalog/unpublish`,
+                { credentials : "include", method : "DELETE", body : JSON.stringify({ "id" : filter?.id }) },
+                () => {
+                  updateProgress({ showProgress : false });
+                  enqueueSnackbar("Filter unpublished", {
```
Comment (leecalcote) [inline ui/components/Filters.js:310]: What filter? Provide the name as confirmation to the user.

@Chadha93, remember when you had to go through the code base and make the notifications uniform by adding the name of the thing into the toaster message... ? Please uphold this convention.

File: ui/components/Filters.js [hunk 6]
```diff
@@ -224,9 +234,39 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
     CLONE_FILTERS : {
       name : "CLONE_FILTER",
       error_msg : "Failed to clone filter file"
+    },
+    PUBLISH_CATALOG : {
+      name : "PUBLISH_CATALOG",
```
Comment (leecalcote) [inline ui/components/Filters.js:239]: This name is a bit of a misnomer, no? Please rename accordingly.

File: ui/components/Filters.js [hunk 7]
```diff
@@ -241,6 +281,58 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
...
+    }
+  };
+
+  const handleUnpublishModal = (ev, filter) => {
+    if (canPublishFilter) {
+      ev.stopPropagation();
+      return async () => {
+        let response = await modalRef.current.show({
+          title : `Unpublish Catalog item?`,
+          subtitle : `Are you sure you want to unpublish ${filter?.name}?`,
```
Comment (leecalcote) [inline ui/components/Filters.js:300]: Yay!

Comment (leecalcote) [inline ui/components/Filters.js:300]:
```suggestion
          subtitle : `Are you sure that you want to unpublish "${filter?.name}"?`,
```

File: ui/components/Filters.js [hunk 8]
```diff
@@ -409,6 +501,29 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
...
   };
 
+  const handlePublish = (catalog_data) => {
+    updateProgress({ showProgress : true });
+    dataFetch(
+      `/api/filter/catalog/publish`,
+      { credentials : "include", method : "POST", body : JSON.stringify(catalog_data) },
+      () => {
+        updateProgress({ showProgress : false });
+        enqueueSnackbar("Filter Published!", {
```
Comment (leecalcote) [inline ui/components/Filters.js:511]: Why are you capitalizing "Published"? Please don't. 

Also, please don't yell at the user! It's not nice, is it?

File: ui/components/Filters.js [hunk 9]
```diff
@@ -575,6 +690,27 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
     }
   }
 
+  const handleDownload = (e, id, name) => {
+    e.stopPropagation();
+    updateProgress({ showProgress : true });
+    try {
+      downloadFile({ id, name, type : "filter" })
+      updateProgress({ showProgress : false });
+      enqueueSnackbar(`"${name}" Filter downloaded`, {
```
Comment (leecalcote) [inline ui/components/Filters.js:699]: What filter?

File: ui/components/Filters.js [hunk 10]
```diff
@@ -575,6 +690,27 @@ function MesheryFilters({ updateProgress, enqueueSnackbar, closeSnackbar, user,
...
+        action : function Action(key) {
+          return (
+            <IconButton key="close" aria-label="Close" color="inherit" onClick={() => closeSnackbar(key)}>
+              <CloseIcon />
+            </IconButton>
+          );
+        }
+      });
+    } catch (e) {
+      console.error(e);
```
Comment (leecalcote) [inline ui/components/Filters.js:710]: What are we doing when this happens?

File: ui/utils/fileDownloader.js (hunk 1 above)
Comment (leecalcote) [reply ui/utils/fileDownloader.js:15]: Sounds appropriate to me. @aabidsofi19 will you jump in with a code suggestion (make a commit to this PR/branch)?

**Review by leecalcote — CHANGES_REQUESTED** [review]

**Comment by leecalcote** [comment]
Urgently needed. Merging. Please follow up on comments.

**Comment by Chadha93** [comment]
> Urgently needed. Merging. Please follow up on comments.

Working on comments, I just checked the publish is breaking for filters, as Publish modal was not updated.

**Pull request merged**